okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimitingInterceptor.java

```diff
@@ -0,0 +1,163 @@
+package com.palantir.remoting3.okhttp;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.netflix.concurrency.limits.Limiter;
+import java.io.IOException;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import okhttp3.Call;
+import okhttp3.Callback;
+import okhttp3.Interceptor;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.ResponseBody;
+import okio.AsyncTimeout;
+import okio.BufferedSource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * WIP docs for benefit of reviewer.
+ *
+ * An interceptor for limiting the concurrency of requests to an endpoint.
+ *
+ * Requests must be tagged (before reaching this point) with a ConcurrencyLimitTag. At this point, we block on
+ * receiving a permit to run the request, and store the listener in the tag.
```

Comment:
I think it would be better to just use the QosHandler directly in here rather than passing around this tag and requiring it to be set